mmseg/models/backbones/swin_transformer.py

```diff
@@ -0,0 +1,657 @@
+from copy import deepcopy
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import build_norm_layer, trunc_normal_init
+from mmcv.cnn.bricks.registry import ATTENTION
+from mmcv.cnn.bricks.transformer import FFN, build_dropout
+from mmcv.runner.base_module import BaseModule, ModuleList
+from torch.nn.modules.utils import _pair as to_2tuple
+
+from ..builder import BACKBONES
+from ..utils import PatchEmbed
+
+
+class PatchMerging(BaseModule):
+    """Merge patch feature map.
+
+    This layer use nn.Unfold to group feature map by kernel_size, and use norm
+    and linear layer to embed grouped feature map.
+    Args:
+        input_resolution (tuple): The size of input patch resolution.
+        in_channels (int): The num of input channels.
+        out_channels (int): The num of output channels.
+        kernel_size (int | tuple, optional): the kernel size in the unfold
+            layer. Defaults: 2.
+        stride (int | tuple, optional): the stride of the sliding blocks in the
+            unfold layer.
+            Defaults: None. (Default to be equal with kernel_size).
+        padding (int | tuple, optional): zero padding width in the unfold
+            layer. Defaults: 0.
+        dilation (int | tuple, optional): dilation parameter in the unfold
+            layer. Defaults: 1.
+        bias (bool, optional): Whether to add bias in linear layer or not.
+            Defaults: False.
+        norm_cfg (dict, optional): Config dict for normalization layer.
+            Defaults: dict(type='LN').
+        init_cfg (dict, optional): The extra config for initialization.
+            Defaults: None.
+    """
+
+    def __init__(self,
+                 input_resolution,
+                 in_channels,
+                 out_channels,
+                 kernel_size=2,
+                 stride=None,
+                 padding=0,
+                 dilation=1,
+                 bias=False,
+                 norm_cfg=dict(type='LN'),
+                 init_cfg=None):
+        super().__init__(init_cfg)
+        H, W = input_resolution
+        self.input_resolution = input_resolution
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+
+        if stride is None:
+            stride = kernel_size
+        kernel_size = to_2tuple(kernel_size)
+        stride = to_2tuple(stride)
+        padding = to_2tuple(padding)
+        dilation = to_2tuple(dilation)
+        self.sampler = nn.Unfold(kernel_size, dilation, padding, stride)
```

Comment:
Check if https://github.com/SwinTransformer/Swin-Transformer-Semantic-Segmentation/blob/87e6f90577435c94f3e92c7db1d36edc234d91f6/mmseg/models/backbones/swin_transformer.py#L283-L286 is also correctly handled.

Reply:
The padding may need to calculate by users.